### ISSUE TITLE
gui: Should warn when adding a folder that is a parent of an existing folder

### ISSUE THREAD

**Issue by seansimonanimation** [syncthing/syncthing#3197]
Hi,

I work with a team of animators, and our Syncthing share has a huge, nested share structure that currently has about 80GB in it, and it's growing a little bit every day. 
I don't know how difficult this would be to make happen (I imagine very), but there have been a couple times in working with Syncthing with a team that I have had to deal with infinite directory loops. Oftentimes they are a result from erring on some Syncthing share algebra, but the fact that it's happened twice in the past couple months could be cause for concern. In those instances, a popup warning that lets us know something along the lines of "Continuing with what you are doing could result in an infinite directory loop!" would have been really nice to keep us from making a that terrible mistake.

If you can't pull the directory structure at share creation, then maybe while it's starting its initial sync, run some pattern recognition that kills the sync process for the affected shares? This might help a lot since not all of my people are very technically inclined. Just a thought. Thanks!

Syncthing Version: v0.13.0
OS Version: Windows 7,10

**Comment by AudriusButkevicius** [comment]
What do you mean infinite loops? Syncthing does not follow symlinks so these should be impossible, unless you are a magican

**Comment by seansimonanimation** [comment]
The first time it happened, it was because one of my team members got confused about her own directory setup, so she put one share inside another share that she wasn't supposed to, causing it to reference itself over and over again.

The second time, it happened just now. I just had to clean up someone else's share that looked like *share root*\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\resources\

I don't know how it happened, but it did.

**Comment by AudriusButkevicius** [comment]
So none of these seem to be infinite as you claim so i am not sure what you want us to do?

**Comment by calmh** [comment]
I guess this can happen if you have two folders, A and B, and on one device you nest A inside of B and on the other side you nest B inside of A. *Don't do that*, though. :) We already have a warning against nested folders, as it's a weird config that may have unforeseen consequences. I think this is one of them and covered by the existing warning.

**Comment by sisu4syncthing** [comment]
@seansimonanimation 
I would like to see the folder config(s) of the folder(s), which had the loops.
That could be helpful for others to prevent such unwanted behaviour...

**Comment by seansimonanimation** [comment]
Every time I've seen it, I have fixed the issue by deleting and remaking the shares on the other end. I'll try to recreate the situation.

**Comment by AudriusButkevicius** [comment]
I am closing this for now as we already warn about nested folders.

**Comment by seansimonanimation** [comment]
@sisu4syncthing 
@AudriusButkevicius 
Hey, it happened again.
So, here's the configuration, and I'm typing this from the culprit system.
So our project structure starts at the folder C:\Users\<user>\Documents\<root>
The culprit share, let's call it project "cartoons", is supposed to be at C:\Users\<user>\Documents\<root>\Projects\cartoons\
A nested structure makes sense for us because it keeps things organized, and allows us to keep things segmented between our team members, for instance, you don't want someone working on project 1 to have the files from project 2. So, we use nested shares, we're usually just very careful about it. It works for us.

The problem arises when you have someone who needs to have multiple shares, say... a project share, as well as their own personal share space. This is when file organizational skills are important. What happened this time is that we created a new share ABOVE an existing share.

The culprit system's "cartoons" share is at C:\Users\<user>\
Where it is supposed to be at C:\Users\<user>\Documents\<root>\Projects\cartoons\
Again, the root share is at C:\Users\<user>\Documents\<root>

This resulted in the entire user folder going inside the Cartoon folder. Since the user folder contains the root, the root went inside there, and the cartoons folder is inside the root so cartoons got made inside there, which contains root, and so on and so forth. Hence, infinite loop. It was probably created by accident, or Syncthing moved the share somehow (which I doubt). Either way, I don't remember there being a warning.

So yeah, I hope this helped. If you need anything else from me, please let me know. Thanks

**Comment by calmh** [comment]
> So, we use nested shares, we're usually just very careful about it. It works for us.

We're supposed to warn about this in the UI. However we might not do that correctly when a new share is above rather than inside an existing one. That's a bug in that case, so in reopening to investigate. 

Note though that I get the impression that you may be ignoring the nested folder warning anyway and have become desentized to it. But that warning is the warning against infinite folder recursion. :)

**Comment by sisu4syncthing** [comment]
I can't remember the exact constellation of my folders right now, but I remember that I missed that "nested folder message" too, when I was testing some time/versions ago.
Luckily, I never had the infinite folders recursion, although I tried to produce them with strange and ill folder settings.

**Comment by calmh** [comment]
Yeah. Having a folder in `tt/s2` already, I get this when adding something underneat:

<img width="940" alt="screen shot 2016-07-08 at 08 18 20" src="https://cloud.githubusercontent.com/assets/125426/16679012/c063857c-44e4-11e6-99dc-2a97d830dc90.png">

but adding something above is not complained about:

<img width="981" alt="screen shot 2016-07-08 at 08 18 36" src="https://cloud.githubusercontent.com/assets/125426/16679020/d4e1888c-44e4-11e6-9614-dd6d149dc560.png">

**Comment by tomachinz** [comment]
<img width="809" alt="infinite folder recursive loop bug not feature" src="https://cloud.githubusercontent.com/assets/1026291/21510267/437f1842-ccf6-11e6-8a78-521d4483073c.png">

My setup is not complex. Getting started with Syncthing planning to roll it out as replacement for a 50 GB dropbox folder. However, this infinite loop issue I got reared it's head very early on - during my very first experiment sharing a folder with a Noob.

I've been studying operating system theory my entire life and have used just about every folder sync tool there is:

- carbon copy cloner
- dropbox / drive / icloud / live
- bittorrent sync (was awesome until the upgrade!)
- rsync / grsync / Synkron
- Dupe Guru
- crashplan / time machine

So far best was Bittorrent Sync, and I would have to say Carbon Copy Cloner (but its one way sync only there most of the time, bi-directional works pretty good though to be fair).

But this issue with Syncthing - not sure I can figure this out? I'm stuck as fuck. Maybe I'll remove and re-add the folder one day - not sure if this is possible it might be caused on my fried Marlene's computer because I dicked around on that in order to *create* the problem.
